### PR TITLE
Emscripten support

### PR DESCRIPTION
**Pull request by Milek7**
This is currently WIP, but I would appreciate feedback on it.
Libtimidity is restored with mixer output, Emscripten support is added and various fixes for Emscripten quirks are applied.
Dockerfile for build environment is here: https://github.com/Milek7/openttd-emscripten-dockerfile/tree/sdl1

It currently works on SDL1 on 32bpp blitter (8bpp blitter doesn't work, I suspect broken SDL palette support in emscripten shim). It doesn't look correct though, because html provides BGRA buffer and all blitters assume RGBA. Modifying `union Colour` works, but this is hacky solution.

It works too on SDL2 branch, with significantly better performance, both on 8bpp (with correct colors) and 32bpp (same BGRA issue, but it is faster than 8bpp). Maybe this PR should wait on SDL2, as it works much better on it.

Network is handled by emscripten as WebSockets, both UDP and TCP. It requires proxy, like this one: https://gist.github.com/Milek7/b921ec9f8d875fe2d5a8b06bfd533834
sockaddr sizes and NetworkAddress comparing needs to be investigated, current workaround is to delete domain name after resolving, and comparing generated hostnames by strcmp. Using original code caused duplicated entries in gamelist as UDP queries received weren't matched to correct entry.

### CODE REVIEW

**Comment by glx22** [comment]
It's hidden in the commit-checker log but
```
*** b/src/video/sdl_v.h:56: Use tabs for indentation: '    uint32 cur_ticks;'
*** b/src/video/sdl_v.h:57: Use tabs for indentation: '    uint32 last_cur_ticks;'
*** b/src/video/sdl_v.h:58: Use tabs for indentation: '    uint32 next_tick;'
*** b/src/video/sdl_v.h:59: Use tabs for indentation: '    std::thread draw_thread;'
*** b/src/video/sdl_v.h:60: Use tabs for indentation: '    std::unique_lock<std::recursive_mutex> draw_lock;'
```

File: src/network/core/address.cpp [hunk 1]
```diff
@@ -242,9 +242,16 @@ SOCKET NetworkAddress::Resolve(int family, int socktype, int flags, SocketList *
...
+	hints.ai_flags &= ~AI_ADDRCONFIG;
+#endif
+
 	int e = getaddrinfo(StrEmpty(this->hostname) ? nullptr : this->hostname, port_name, &hints, &ai);
 
-	if (reset_hostname) strecpy(this->hostname, "", lastof(this->hostname));
+#ifndef __EMSCRIPTEN__
+	if (reset_hostname)
+#endif
+		strecpy(this->hostname, "", lastof(this->hostname));
```
Comment (glx22) [inline src/network/core/address.cpp]: This is not really following the coding style. Can't you force `reset_hostname` to `true` for Emscripten instead ?

File: src/network/core/address.cpp [hunk 2]
```diff
@@ -301,7 +306,12 @@ static SOCKET ConnectLoopProc(addrinfo *runp)
 
 	if (!SetNoDelay(sock)) DEBUG(net, 1, "[%s] setting TCP_NODELAY failed", type);
 
-	if (connect(sock, runp->ai_addr, (int)runp->ai_addrlen) != 0) {
+	int err = connect(sock, runp->ai_addr, (int)runp->ai_addrlen);
+#ifndef __EMSCRIPTEN__
+	if (err != 0) {
+#else
+	if (err != 0 && errno != EINPROGRESS) {
```
Comment (LordAro) [inline src/network/core/address.cpp:315]: Presumably this is because Emscripten doesn't define `EINPROGRESS` ? Might be cleaner to define it to something (0 would work?) than the 2 separate if statements

Also add a comment explaining why it's necessary

File: src/network/core/address.cpp [hunk 3]
```diff
@@ -301,7 +306,12 @@ static SOCKET ConnectLoopProc(addrinfo *runp)
...
+#endif
```
Comment (LordAro) [inline src/network/core/address.cpp]: missing `/* __EMSCRIPTEN__ */` suffix

also in other places

File: src/network/core/address.cpp [hunk 4]
```diff
@@ -242,6 +242,11 @@ SOCKET NetworkAddress::Resolve(int family, int socktype, int flags, SocketList *
 		strecpy(this->hostname, fam == AF_INET ? "0.0.0.0" : "::", lastof(this->hostname));
 	}
 
+#ifdef __EMSCRIPTEN__
```
Comment (LordAro) [inline src/network/core/address.cpp:246]: would appreciate comment about why this is necessary

File: src/network/core/address.cpp [hunk 5]
```diff
@@ -24,8 +24,8 @@
 const char *NetworkAddress::GetHostname()
 {
 	if (StrEmpty(this->hostname) && this->address.ss_family != AF_UNSPEC) {
-		assert(this->address_length != 0);
-		getnameinfo((struct sockaddr *)&this->address, this->address_length, this->hostname, sizeof(this->hostname), nullptr, 0, NI_NUMERICHOST);
+		int size = this->address.ss_family == AF_INET6 ? sizeof(struct sockaddr_in6) : sizeof(struct sockaddr_in);
```
Comment (LordAro) [inline src/network/core/address.cpp:28]: why this change? seems like it should be explained sufficiently and split into its own commit
(and this->address_length removed?)

File: src/network/core/address.h [hunk 6]
```diff
@@ -138,11 +138,15 @@ class NetworkAddress {
...
 	int CompareTo(NetworkAddress &address)
 	{
+#ifndef __EMSCRIPTEN__
 		int r = this->GetAddressLength() - address.GetAddressLength();
 		if (r == 0) r = this->address.ss_family - address.address.ss_family;
 		if (r == 0) r = memcmp(&this->address, &address.address, this->address_length);
 		if (r == 0) r = this->GetPort() - address.GetPort();
 		return r;
+#else
+		return strcmp(GetHostname(), address.GetHostname());
```
Comment (LordAro) [inline src/network/core/address.h]: explanation please.

also it would be easier to read if the ifndef was inverted, imo

File: src/network/core/os_abstraction.h [hunk 7]
```diff
@@ -145,12 +145,16 @@ typedef unsigned long in_addr_t;
...
 {
+#ifndef __EMSCRIPTEN__
 #ifdef _WIN32
 	u_long nonblocking = 1;
 #else
 	int nonblocking = 1;
 #endif
 	return ioctlsocket(d, FIONBIO, &nonblocking) == 0;
+#else
+	return true;
```
Comment (LordAro) [inline src/network/core/os_abstraction.h]: could roll this into the existing ifdef with an `#elif`, i think?

File: src/network/core/tcp.cpp [hunk 8]
```diff
@@ -236,7 +236,13 @@ bool NetworkTCPSocketHandler::CanSendReceive()
 	FD_SET(this->sock, &write_fd);
 
 	tv.tv_sec = tv.tv_usec = 0; // don't block at all.
-	if (select(FD_SETSIZE, &read_fd, &write_fd, nullptr, &tv) < 0) return false;
+
+#ifndef __EMSCRIPTEN__
+	int nfds = FD_SETSIZE;
```
Comment (LordAro) [inline src/network/core/tcp.cpp:241]: again, if this is because EMSCRIPTEN doesn't define `FD_SETSIZE`, it should be set separately, instead of creating an extra variable

File: src/video/sdl_v.cpp [hunk 9]
```diff
@@ -74,7 +74,8 @@ static void UpdatePalette(bool init = false)
 		pal[i].unused = 0;
 	}
 
-	SDL_SetColors(_sdl_screen, pal, _local_palette.first_dirty, _local_palette.count_dirty);
+	if (BlitterFactory::GetCurrentBlitter()->GetScreenDepth() == 8)
```
Comment (LordAro) [inline src/video/sdl_v.cpp]: codestyle - if statements with a single statement as a body need to be on the same line, or enclosed by braces. Probably want to add the braces

& other places

File: src/video/sdl_v.cpp [hunk 10]
```diff
@@ -656,55 +657,27 @@ void VideoDriver_SDL::Stop()
...
+	drv->LoopOnce();
+}
+#endif
 
-	DEBUG(driver, 1, "SDL: using %sthreads", _draw_threaded ? "" : "no ");
+void VideoDriver_SDL::LoopOnce()
+{
+	    uint32 mod;
+	    int numkeys;
+	    Uint8 *keys;
```
Comment (LordAro) [inline src/video/sdl_v.cpp]: weird indentation going on here

File: src/video/sdl_v.cpp [hunk 11]
```diff
@@ -656,55 +657,27 @@ void VideoDriver_SDL::Stop()
 	}
 }
 
-void VideoDriver_SDL::MainLoop()
```
Comment (LordAro) [inline src/video/sdl_v.cpp:647]: This code shouldn't be moved around while modifying it - difficult to tell what's been changed - should be split into separate commits

File: src/video/sdl_v.h [hunk 12]
```diff
@@ -14,8 +14,16 @@
 
 #include "video_driver.hpp"
 
+#ifdef __EMSCRIPTEN__
+void em_loop(void *arg);
```
Comment (LordAro) [inline src/video/sdl_v.h:18]: is there a particular reason not to make this a (conditional) class method? would save the duplicate `friend` definition

File: src/video/sdl_v.h [hunk 13]
```diff
@@ -42,6 +52,12 @@ class VideoDriver_SDL : public VideoDriver {
 	int PollEvent();
 	bool CreateMainSurface(uint w, uint h);
 	void SetupKeyboard();
+
+	uint32 cur_ticks;
```
Comment (LordAro) [inline src/video/sdl_v.h:56]: I don't like that these are necessary at all...

File: src/3rdparty/squirrel/squirrel/sqvm.cpp [hunk 14]
```diff
@@ -6,7 +6,10 @@
 
 #include <squirrel.h>
 #include "sqpcheader.h"
+#pragma push_macro("type")
```
Comment (LordAro) [inline src/3rdparty/squirrel/squirrel/sqvm.cpp:10]: comment explaination for this please

**Review by LordAro — DISMISSED** [review]

File: src/network/core/address.cpp (hunk 2 above)
Comment (Milek7) [reply src/network/core/address.cpp:315]: Emscripten doesn't support synchronous sockets, so they are always created nonblocking. (thus connect always returns EINPROGRESS, and that cannot be interpreted as error)

File: src/network/core/address.cpp (hunk 5 above)
Comment (Milek7) [reply src/network/core/address.cpp:28]: Emscripten getnameinfo returns error when passed size other than sizeof(sockaddr_in)
https://github.com/emscripten-core/emscripten/issues/7636
It isn't guarded by emscripten specific ifdef as it shouldn't hurt other platforms.

File: src/network/core/address.cpp (hunk 4 above)
Comment (Milek7) [reply src/network/core/address.cpp:246]: emscripten getaddrinfo fails when given AI_ADDRCONFIG.
reset_hostname is set because comparing addresses doesn't match entries properly, thus workaround is to compare strings returned from getnameinfo

File: src/network/core/os_abstraction.h (hunk 7 above)
Comment (Milek7) [reply src/network/core/os_abstraction.h]: I don't see how, existing ifdef ends before ioctlsocket. (early return maybe, but that could cause unreachable code warnings)

File: src/network/core/tcp.cpp (hunk 8 above)
Comment (Milek7) [reply src/network/core/tcp.cpp:241]: It does define it to 1024, but its `select` support only 64
https://github.com/emscripten-core/emscripten/issues/1711

File: src/video/sdl_v.h (hunk 12 above)
Comment (Milek7) [reply src/video/sdl_v.h:18]: It is called by function pointer by emscripten. Is it safe to assume that calling convention would match?

File: src/network/core/tcp.cpp (hunk 8 above)
Comment (LordAro) [reply src/network/core/tcp.cpp:241]: hrm, well just a significant comment explaining why then, probably with a link to the bug

File: src/network/core/address.cpp (hunk 2 above)
Comment (LordAro) [reply src/network/core/address.cpp:315]: Fair enough, just a comment explaining as such then

**Comment by Milek7** [comment]
waiting for #7086 to merge both sdl1 and sdl2 emscripten changes at same time

**Comment by Milek7** [comment]
Superseded by #8355
I might create separate PR for libtimidity later.
